client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -310,7 +310,7 @@ where
 		startup_point: &StartupPoint<Block>,
 	) -> Result<Vec<FollowEvent<Block::Hash>>, SubscriptionManagementError> {
 		// The block was already pinned by the initial block events or by the finalized event.
-		if !self.sub_handle.pin_block(notification.hash)? {
+		if !self.sub_handle.pin_block(&self.sub_id, notification.hash)? {
```

Comment:
Yep, we keep track internally of the calls made to the `pin_block`.
The second call will return `false` to signal we generated the events for this block. 

The `pin_block` can happen from the
- import blocks 
- finalized blocks

We use the `BlockStateMachine` to ensure that we untrack a block from memory only when we receive two calls to the `pin_block` and one `unpin_block` . 

This will protect us against:
- import event comes in: we generate events
- user calls unpin: we previously lost track of the block
- finalized event comes in: we previously regenerated events

However if both events come in (import and finalized) and the user calls unpin; _and_ the backend gives us the same block again hash again (most probably a bug), we do not protect against this :D